### PR TITLE
msg: client bind

### PR DESCRIPTION
**Pull request by yuyuyu101**

### CODE REVIEW

**Comment by yuyuyu101** [comment]
refactor from https://github.com/ceph/ceph/pull/7256

**Comment by yuyuyu101** [comment]
we also met this problem when public and cluster are different net but can be interconnect. it will cause mon electing always and osd false down

File: src/msg/async/AsyncConnection.cc [hunk 1]
```diff
@@ -859,6 +859,18 @@ ssize_t AsyncConnection::_process_connection()
           cs.close();
         }
 
+        {
+          entity_addr_t addr2bind = msgr->get_myaddr();
+          if (msgr->cct->ms_bind_before_connect && (!addr2bind.is_blank_ip())) {
```
Comment (liewegas) [inline src/msg/async/AsyncConnection.cc]: get_myaddr() returns by reference, right?  can use it directly instead of copying the value to addr2bind

Comment (liewegas) [inline src/msg/async/AsyncConnection.cc]: eh, i guess if we have this option enabled by default it doesn't matter

**Comment by liewegas** [comment]
<pre>/home/jenkins-build/build/workspace/ceph-pull-requests/src/msg/simple/Pipe.cc: In member function ‘int Pipe::accept()’:
/home/jenkins-build/build/workspace/ceph-pull-requests/src/msg/simple/Pipe.cc:361:20: error: ‘class CephContext’ has no member named ‘ms_bind_before_connect’
     if (msgr->cct->ms_bind_before_connect && (!addr2bind.is_blank_ip())) {
                    ^
</pre>

File: src/msg/async/AsyncConnection.cc (hunk 1 above)
Comment (yuyuyu101) [inline src/msg/async/AsyncConnection.cc]: we will modify addr2bind, so nonreference is suitable

**Comment by yuyuyu101** [comment]
retest this please

File: src/ceph_osd.cc [hunk 2]
```diff
@@ -563,6 +573,9 @@ int main(int argc, const char **argv)
   r = ms_hb_front_server->bind(hb_front_addr);
   if (r < 0)
     exit(1);
+  r = ms_hb_front_client->client_bind(hb_back_addr);
```
Comment (yuyuyu101) [inline src/ceph_osd.cc]: @yuriw sorry, I have a mistake. you may need to update this pr to branch and schedule qa

**Comment by yuriw** [comment]
http://pulpito.front.sepia.ceph.com/yuriw-2017-01-17_18:50:47-rados-wip-yuri-testing_2017_1_18-distro-basic-smithi/

http://pulpito.front.sepia.ceph.com/yuriw-2017-01-19_16:12:41-rados-wip-yuri-testing_2017_1_18---basic-smithi/

and retry

http://pulpito.ceph.com/yuriw-2017-01-19_23:03:46-rados-wip-yuri-testing_2017_1_18---basic-smithi/

compare to master

http://pulpito.ceph.com/teuthology-2017-01-20_18:52:46-rados-master---basic-smithi/

File: src/msg/simple/SimpleMessenger.cc [hunk 3]
```diff
@@ -306,6 +306,27 @@ int SimpleMessenger::rebind(const set<int>& avoid_ports)
   return accepter.rebind(avoid_ports);
 }
 
+
+int SimpleMessenger::client_bind(const entity_addr_t &bind_addr)
+{
+  lock.Lock();
+  if (did_bind) {
+    assert(my_inst.addr == bind_addr);
```
Comment (Liuchang0812) [inline src/msg/simple/SimpleMessenger.cc]: seems that miss `lock.Unlock()` ?

Comment (yuyuyu101) [inline src/msg/simple/SimpleMessenger.cc]: good catch

**Comment by dillaman** [comment]
It looks like this will break librados functionality when used behind a NAT since it learns the address of the gateway and then attempts to bind using it instead of its local IP address [1].

[1] http://tracker.ceph.com/issues/20049

**Pull request merged**